File: apps/web/src/api/coupleHonest.ts
/**
 * Ehrliches Mitteilen — der einzige Bereich des Paarraums ohne KI.
 *
 * Kein Endpunkt hier ruft Echo auf. Was zwei Menschen einander mitteilen, verlässt den
 * Server nicht: kein Modellaufruf, kein Transfer, keine Deutung. Die Sicherheitsschicht
 * läuft trotzdem, aber mit dem deterministischen Stichwort-Boden, der lokal rechnet.
 */
import { apiClient } from './client'

/** „gehoert" = lies erst · „gegenueber" = die andere ist dran · null = du bist dran. */
export type HonestBlock = 'gehoert' | 'gegenueber' | null

export interface HonestShare {
  id: string
  is_own: boolean
  name: string
  impulse: string | null
  impulse_label: string | null
  body: string
  heard: boolean
  /** Wie es angekommen ist – sichtbar für beide, das ist der Sinn der Quittung. */
  heard_as: string | null
  heard_as_label: string | null
  created_at: string
  /** Nur am eigenen Beitrag gesetzt – nie am fremden. */
  safety: { level?: string; source?: string } | null
}

export interface HonestRound {
  id: string
  status: 'arriving' | 'open' | 'closed'
  created_at: string
  closed_at: string | null
}

export interface HonestHistoryEntry {
  id: string
  created_at: string
  closed_at: string | null
  share_count: number
}

export interface HonestView {
  round: HonestRound | null
  arrival_own: { body: string; safety: Record<string, unknown> | null } | null
  arrival_other: { body: string; name: string | null } | null
  /** Dass sie da ist, sieht man immer. Was sie schrieb, erst wenn beide da sind. */
  arrival_other_done: boolean
  shares: HonestShare[]
  my_turn: boolean
  blocked_reason: HonestBlock
  /** Schlüssel → Frage + Schreibhilfe. Die Hilfe gehört zur Methode, nicht zum Layout. */
  impulses: Record<string, { label: string; hint: string }>
  /** Die drei festen Quittungen. Kein Freitext – sonst würde daraus eine Antwort. */
  acknowledgements: Record<string, string>
  /** Die wievielte Runde – damit sich die Übung wie eine Übung anfühlt. */
  round_number: number
  partner_name: string | null
  history: HonestHistoryEntry[]
  /** Sicherheitshinweis – nur an die schreibende Person, nur beim Absenden. */
  notice: string | null
}

export interface HonestClosedRound {
  round: HonestRound
  shares: { id: string; is_own: boolean; name: string; impulse_label: string | null
            body: string; created_at: string }[]
}

const basis = (coupleId: string) => `/couple/links/${coupleId}/mitteilen`

export const coupleHonestApi = {
  get: (coupleId: string) =>
    apiClient.get<HonestView>(basis(coupleId)).then(r => r.data),

  begin: (coupleId: string) =>
    apiClient.post<HonestView>(`${basis(coupleId)}/beginnen`).then(r => r.data),

  arrive: (coupleId: string, body: string) =>
    apiClient.post<HonestView>(`${basis(coupleId)}/ankommen`, { body }).then(r => r.data),

  share: (coupleId: string, body: string, impulse: string | null) =>
    apiClient.post<HonestView>(`${basis(coupleId)}/beitrag`, { body, impulse })
      .then(r => r.data),

  markHeard: (coupleId: string, shareId: string, kind: string) =>
    apiClient.post<HonestView>(`${basis(coupleId)}/beitraege/${shareId}/gehoert`, { kind })
      .then(r => r.data),

  close: (coupleId: string) =>
    apiClient.post<HonestView>(`${basis(coupleId)}/abschliessen`).then(r => r.data),

  readRound: (coupleId: string, roundId: string) =>
    apiClient.get<HonestClosedRound>(`${basis(coupleId)}/runden/${roundId}`)
      .then(r => r.data),
}

File: services/api/app/schemas/couple_honest.py
"""Schemas: Ehrliches Mitteilen."""
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class HonestArrive(BaseModel):
    #: Ein Satz, kein Absatz – die Länge ist Teil der Übung.
    body: str = Field(..., max_length=300)


class HonestShare(BaseModel):
    body: str = Field(..., max_length=1500)
    #: Schlüssel aus dem Impuls-Katalog, oder None bei freiem Text. Beides ist richtig.
    impulse: str | None = None


class HonestHeard(BaseModel):
    #: Eine von drei festen Aussagen: gehoert | beruehrt | schwer. Kein Freitext -
    #: sonst wuerde aus der Quittung eine Antwort, und genau die soll es nicht geben.
    kind: str = "gehoert"


class HonestRound(BaseModel):
    id: UUID
    #: arriving | open | closed
    status: str
    created_at: datetime
    closed_at: datetime | None = None


class HonestArrivalOwn(BaseModel):
    body: str
    #: Nur am eigenen Beitrag gesetzt – nie an dem der anderen Person.
    safety: dict | None = None


class HonestArrivalOther(BaseModel):
    body: str
    name: str | None = None


class HonestShareView(BaseModel):
    id: UUID
    is_own: bool
    name: str
    impulse: str | None = None
    impulse_label: str | None = None
    body: str
    heard: bool
    #: Wie es angekommen ist – sichtbar für beide, das ist der Sinn der Quittung.
    heard_as: str | None = None
    heard_as_label: str | None = None
    created_at: datetime
    safety: dict | None = None


class HonestHistoryEntry(BaseModel):
    id: UUID
    created_at: datetime
    closed_at: datetime | None = None
    share_count: int


class HonestRoundView(BaseModel):
    round: HonestRound | None = None
    arrival_own: HonestArrivalOwn | None = None
    arrival_other: HonestArrivalOther | None = None
    #: Ist die andere Person schon angekommen? Sichtbar wird ihr Satz erst, wenn beide da sind.
    arrival_other_done: bool = False
    shares: list[HonestShareView] = []
    my_turn: bool = False
    #: "gehoert" = lies erst · "gegenueber" = die andere ist dran · None = du bist dran
    blocked_reason: str | None = None
    #: Schlüssel → {label, hint}. Die Schreibhilfe steht im Katalog, nicht in der
    #: Oberfläche – sie gehört zur Methode, nicht zum Layout.
    impulses: dict[str, dict[str, str]] = {}
    acknowledgements: dict[str, str] = {}
    #: Die wievielte Runde – damit sich die Übung wie eine Übung anfühlt.
    round_number: int = 1
    partner_name: str | None = None
    history: list[HonestHistoryEntry] = []
    #: Sicherheitshinweis an die SCHREIBENDE Person, sofern der Stichwort-Boden anschlug.
    notice: str | None = None
